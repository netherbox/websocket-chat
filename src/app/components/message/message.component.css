:host {
  display: block;
  position: relative;
  padding: 12px 23px;
}

:host:hover {
  background: #f2f2f2;
}

:host:hover .tools {
  display: block;
}

.author {
  font-weight: 700;
}

.stamp {
  color: #9a9fa8;
  font-weight: normal;
}

.text {
  margin-top: 5px;
}

button,
input[type="submit"] {
  padding: 5px 10px;
  border: 1px solid #868d98;
  border-radius: 3px;
}

.tools {
  display: none;
  position: absolute;
  top: 12px;
  right: 23px;
}

.tools > * {
  margin-left: 5px;
}

.update .actions > * {
  margin-right: 5px;
}

.update input[type="text"] {
  display: block;
  padding: 12px 14px;
  border: 1px solid #868d98;
  margin: 16px 0 8px;
  border-radius: 3px;
}

.deleted {
  margin-top: 5px;
}
