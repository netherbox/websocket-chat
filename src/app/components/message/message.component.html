<div class="author">
  {{ message?.createdBy }}
  <span class="stamp">{{ message?.createdAt | date: "shortTime" }}</span>
</div>

<ng-container *ngIf="!message.deleted && !showUpdateForm">
  <div class="text">
    {{ message.text }}
  </div>

  <div class="tools" *ngIf="editable">
    <button (click)="onUpdateClick()">Edit</button>
    <button (click)="onDeleteClick()">Delete</button>
  </div>

  <div class="updated" *ngIf="message.updated">
    The message was edited by {{ message.updatedBy }} at
    {{ message.createdAt | date: "shortTime" }}
  </div>
</ng-container>

<ng-container *ngIf="showUpdateForm">
  <form class="update" (submit)="onUpdateMessageSubmit(updateMessageInput)">
    <input
      class="input"
      type="text"
      placeholder="Message"
      required
      #updateMessageInput
      [value]="message.text"
    />
    <div class="actions">
      <input type="submit" value="Submit" />
      <button (click)="onUpdateMessageCancel()">Cancel</button>
    </div>
  </form>
</ng-container>

<div class="deleted" *ngIf="message.deleted">
  The message was deleted by {{ message.deletedBy }} at
  {{ message.createdAt | date: "shortTime" }}
</div>
