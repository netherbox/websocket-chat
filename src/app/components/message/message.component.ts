import { ChangeDetectionStrategy, Component, EventEmitter, Input, Output } from '@angular/core';
import { ChatMessage } from '../../models/chat-message.model';

@Component({
  selector: 'app-message',
  templateUrl: './message.component.html',
  styleUrls: ['./message.component.css'],
  changeDetection: ChangeDetectionStrategy.OnPush
})
export class ChatMessageComponent {
  @Input()
  message: ChatMessage = {
    messageId: '',
    text: '',
  };

  @Input()
  editable = false;

  showUpdateForm = false;

  @Output() update: EventEmitter<any> = new EventEmitter();

  @Output() delete: EventEmitter<any> = new EventEmitter();

  onUpdateClick() {
    this.showUpdateForm = true;
  }

  onUpdateMessageSubmit(messageInput: HTMLInputElement): boolean {
    this.update.emit({ ...this.message, text: messageInput.value });
    this.showUpdateForm = false;
    return false;
  }

  onUpdateMessageCancel() {
    this.showUpdateForm = false;
  }

  onDeleteClick() {
    this.delete.emit({ ...this.message });
  }
}
