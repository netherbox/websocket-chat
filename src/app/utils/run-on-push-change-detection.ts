import { ChangeDetectorRef } from '@angular/core';
import { ComponentFixture } from '@angular/core/testing';

/**
 * Force OnPush change detection run
 * https://github.com/angular/angular/issues/12313#issuecomment-528536934
 */
export async function runOnPushChangeDetection<T>(cf: ComponentFixture<T>) {
  const cd = cf.debugElement.injector.get<ChangeDetectorRef>(
    // tslint:disable-next-line:no-any
    ChangeDetectorRef as any
  );
  cd.detectChanges();
  await cf.whenStable();
  return;
}
