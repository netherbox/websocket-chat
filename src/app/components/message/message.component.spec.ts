import { ComponentFixture, TestBed } from '@angular/core/testing';
import { runOnPushChangeDetection } from '../../utils/run-on-push-change-detection';

import { ChatMessage } from '../../models/chat-message.model';
import { ChatMessageComponent } from './message.component';

describe('ChatMessageComponent', () => {
  let component: ChatMessageComponent;
  let fixture: ComponentFixture<ChatMessageComponent>;
  let testMessage: ChatMessage;

  beforeEach(async () => {
    testMessage = {
      messageId: 'test-messageId',
      text: 'test-text',
    };

    await TestBed.configureTestingModule({
      declarations: [ChatMessageComponent],
    }).compileComponents();

    fixture = TestBed.createComponent(ChatMessageComponent);
    component = fixture.componentInstance;

    component.message = { ...testMessage };

    fixture.detectChanges();
  });

  it('should successfully create component with expected html', () => {
    expect(component).toBeTruthy();
    expect(fixture.nativeElement).toMatchSnapshot();
  });

  describe('editable', () => {
    it('should render tools bar if editable true', () => {
      component.editable = true;
      runOnPushChangeDetection(fixture);

      expect(fixture.nativeElement.querySelector('.tools')).toBeTruthy();
      expect(fixture.nativeElement).toMatchSnapshot();
    });

    it('should render tools bar if editable false', () => {
      component.editable = false;
      runOnPushChangeDetection(fixture);

      expect(fixture.nativeElement.querySelector('.tools')).toBeFalsy();
      expect(fixture.nativeElement).toMatchSnapshot();
    });
  });

  describe('onUpdateClick', () => {
    it('should show update form', () => {
      component.onUpdateClick();
      runOnPushChangeDetection(fixture);

      expect(component.showUpdateForm).toBeTruthy();
      expect(fixture.nativeElement.querySelector('.update')).toBeTruthy();
      expect(fixture.nativeElement).toMatchSnapshot();
    });
  });

  describe('onUpdateMessageCancel', () => {
    it('should hide update form', () => {
      component.onUpdateMessageCancel();
      runOnPushChangeDetection(fixture);

      expect(component.showUpdateForm).toBeFalsy();
      expect(fixture.nativeElement.querySelector('.update')).toBeFalsy();
      expect(fixture.nativeElement).toMatchSnapshot();
    });
  });

  describe('onUpdateMessageSubmit', () => {
    it('should emit update event', () => {
      jest.spyOn(component.update, 'emit');

      component.onUpdateMessageSubmit({
        value: 'test-input-value',
      } as unknown as HTMLInputElement);
      runOnPushChangeDetection(fixture);

      expect(component.update.emit).toBeCalledWith({
        ...testMessage,
        text: 'test-input-value',
      });
      expect(component.showUpdateForm).toBeFalsy();
      expect(fixture.nativeElement.querySelector('.update')).toBeFalsy();
      expect(fixture.nativeElement).toMatchSnapshot();
    });
  });

  describe('onDeleteClick', () => {
    it('should emit delete event', () => {
      jest.spyOn(component.delete, 'emit');

      component.onDeleteClick();
      runOnPushChangeDetection(fixture);

      expect(component.delete.emit).toBeCalledWith(testMessage);
    });
  });
});
