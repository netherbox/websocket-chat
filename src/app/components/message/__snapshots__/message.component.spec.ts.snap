// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ChatMessageComponent editable should render tools bar if editable false 1`] = `
<div
  id="root2"
>
  <div
    class="author"
  >
      
    <span
      class="stamp"
    >
      
    </span>
  </div>
  
  
  <div
    class="text"
  >
     test-text 
  </div>
  
  
  
  
  
</div>
`;

exports[`ChatMessageComponent editable should render tools bar if editable true 1`] = `
<div
  id="root1"
>
  <div
    class="author"
  >
      
    <span
      class="stamp"
    >
      
    </span>
  </div>
  
  
  <div
    class="text"
  >
     test-text 
  </div>
  
  <div
    class="tools"
  >
    <button>
      Edit
    </button>
    <button>
      Delete
    </button>
  </div>
  
  
  
  
</div>
`;

exports[`ChatMessageComponent onUpdateClick should show update form 1`] = `
<div
  id="root3"
>
  <div
    class="author"
  >
      
    <span
      class="stamp"
    >
      
    </span>
  </div>
  
  
  
  <form
    class="update"
  >
    <input
      class="input"
      placeholder="Message"
      required=""
      type="text"
    />
    <div
      class="actions"
    >
      <input
        type="submit"
        value="Submit"
      />
      <button>
        Cancel
      </button>
    </div>
  </form>
  
</div>
`;

exports[`ChatMessageComponent onUpdateMessageCancel should hide update form 1`] = `
<div
  id="root4"
>
  <div
    class="author"
  >
      
    <span
      class="stamp"
    >
      
    </span>
  </div>
  
  
  <div
    class="text"
  >
     test-text 
  </div>
  
  
  
  
  
</div>
`;

exports[`ChatMessageComponent onUpdateMessageSubmit should emit update event 1`] = `
<div
  id="root5"
>
  <div
    class="author"
  >
      
    <span
      class="stamp"
    >
      
    </span>
  </div>
  
  
  <div
    class="text"
  >
     test-text 
  </div>
  
  
  
  
  
</div>
`;

exports[`ChatMessageComponent should successfully create component with expected html 1`] = `
<div
  id="root0"
>
  <div
    class="author"
  >
      
    <span
      class="stamp"
    >
      
    </span>
  </div>
  
  
  <div
    class="text"
  >
     test-text 
  </div>
  
  
  
  
  
</div>
`;
